### PR TITLE
chore(ci): add automated dated release workflow

### LINKED ISSUE #220: The project has no releases, no tags and no release notes
Orbit has no releases and no tags. `git ls-remote --tags` returns nothing, and the GitHub releases page is empty.

[docs/roadmap.md](https://github.com/Noveum/orbit/blob/main/docs/roadmap.md) says this is deliberate:

> Orbit ships continuously from `main`. There are no long lived release branches and no backporting, so self-hosted deployments should track `main` or a recent tag.

Continuous shipping is a good decision and this issue does not propose changing it. But that sentence says "or a recent tag", and there are no tags, so the second option does not exist. Right now a self hoster has exactly one choice, which is to deploy whatever `main` happened to be when they cloned, with no way to describe what they are running, no way to know what changed since, and no way to be told that an upgrade needs action from them.

That same roadmap page promises release notes:

> Anything requiring action from someone self-hosting is labelled `breaking change` and called out in the release notes.

There are no release notes to be called out in.

## What to build

**Dated tags, cut automatically.** Continuous shipping does not conflict with tagging. A tag per week, or per merge to `main`, gives every deployment a name. Calendar versioning fits a continuously shipping project better than semantic versioning, because there is no API surface being promised. `2026.08.14` says more to a self hoster than `v0.4.2` does.

**Generated release notes.** From merged pull requests since the last tag, grouped by label. The `area:` labels already exist and are already applied consistently, so the grouping is free. `breaking change` entries go at the top, under their own heading, with the action required spelled out.

**Automation.** A workflow that does this without anybody remembering. [release-please](https://github.com/googleapis/release-please) and [changesets](https://github.com/changesets/changesets) are the two usual choices, and a small custom workflow is also reasonable given how simple the requirement is. Whichever, it goes in `.github/workflows/` and matches the existing style: plain language step names, narrowest workable `permissions`, and actions pinned to a full commit SHA with no trailing version comment, because `bun run check-comments` covers YAML and rejects them.

**A CHANGELOG, or not.** If release notes live on the GitHub releases page, a `CHANGELOG.md` duplicating them is a second source of truth that will drift. Pick one. The releases page is probably enough.

**Upgrade notes for self hosters.** The one thing a self hoster genuinely needs is to know when a release requires them to run a migration or set a new environment variable. [CLAUDE.md](https://github.com/Noveum/orbit/blob/main/CLAUDE.md) says migrations are applied locally against the target database and never by a job in the platform, so a schema change is always manual work for somebody. That has to appear in the notes, prominently, not as a line in a list of forty.

## Why this matters more than it looks

Every issue in this repository about self hosting eventually reduces to "which version are you running", and there is currently no answer to that question. It also affects how the project is perceived: a repository with no releases reads as unfinished to somebody deciding whether to depend on it, however good the code is.

## Related

The Planning parity milestone has an issue about building a releases feature inside Orbit itself. Once that exists, Orbit's own release notes should be written in Orbit and published from it. Doing this one first with a plain workflow is the right order.

## Files

- `.github/workflows/`
- `docs/roadmap.md`, whose "Releases" section needs updating once this is true
- `docs/self-hosting.md`

### PR DESCRIPTION
**Pull request by Andes-indica**
Closes #220

## Summary

- Runs a weekly and manually dispatchable dated release workflow.
- Generates release notes from merged pull requests in an exact first-parent commit range.
- Recovers the earliest unpublished dated tag on the target history before creating a new tag.
- Keeps existing tags immutable and refuses target mismatches.
- Prevents recovered historical tags from overlapping later release-note ranges.
- Parses GitHub API responses with shared Zod validators.
- Documents the release model and self-hosting behavior.

## Verification

- 33 focused workflow and release-note tests pass.
- Repository lint, comment policy, typechecks, unit and integration tests, Playwright, build, migrations, links, and CodeQL are green on the exact head.
- CodeRabbit completed its exact-head review and Greptile reports 5/5 with no blocker.
- All 33 review threads are resolved.
- Vercel completed an exact-SHA Preview through a temporary internal validation ref, and that ref was deleted immediately after success.

### CODE REVIEW

**Comment by Andes-indica** [comment]
even i tried running manually. it worked.

 I just forgot to attach this 
image
  
image

**Review by imshashank — CHANGES_REQUESTED** [review]
Thanks for taking #220 on, and for pinning the actions to full SHAs without being asked. A few things need sorting before this can go in, one of which means the workflow does not currently do what the PR says.

## The since date never reaches the script

The `Determine last tag and since` step computes `since_date` and writes it to `$GITHUB_OUTPUT`, but the Python step only sets `GITHUB_TOKEN` in its `env:` block. Inside the script:

```python
since = os.environ.get('INPUT_SINCE') or os.environ.get('since_date') or os.environ.get('GITHUB_SINCE')
```

None of those three are ever set, so `since` always falls through to the seven day fallback. The whole tag detection step is dead code, and every release ends up with notes covering the last seven days regardless of when the previous tag was cut.

On a `push` trigger that means consecutive releases repeat the same pull requests, which is the opposite of what release notes are for. Pass it explicitly:

```yaml
env:
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  since_date: ${{ steps.since.outputs.since_date }}
```

## Releasing on every push to main is a bigger decision than it looks

`on: push: branches: [main]` cuts a tag and a GitHub release for every merge. We merged three PRs yesterday, so that is three releases in a day, tagged `2026.08.17`, `2026.08.17-1`, `2026.08.17-2`.

#220 asks for releases, tags and release notes, and I read that as wanting something a self hoster can point at and reason about, not a marker per commit. `docs/self-hosting.md` in this PR tells people to "track `main` or a recent dated tag", which only means something if a tag represents a considered boundary.

My preference is the weekly schedule plus `workflow_dispatch`, and dropping the push trigger. If you think per merge is right, argue it and I will listen, but it should be a stated choice rather than a default.

## Cancelling a release halfway

```yaml
concurrency:
  group: automated-release
  cancel-in-progress: true
```

The job pushes a tag and then creates a release in a separate step. Cancelling between those two leaves a tag on the remote with no release attached, and the next run picks a different dated suffix rather than repairing it. Releases want `cancel-in-progress: false` so a second run queues instead of severing the first.

## The release call cannot fail

```bash
curl -sS -H "Authorization: token $GITHUB_TOKEN" ... -d "$data"
```

`-sS` without `-f` means an API error prints a JSON error body and exits 0, so the step goes green having pushed a tag and created nothing. Add `--fail-with-body`, or use `gh release create` which is already on the runner and would replace most of this step.

## Python in a Bun repo

`CLAUDE.md` is firm that Bun is the script runner and that repo tooling lives in `scripts/` as TypeScript. A 60 line Python heredoc inside a workflow is the one place nobody will think to look when it breaks, and it cannot be run locally the way `scripts/` can.

Moving it to `scripts/release-notes.ts` run with `bun` would match the repo, get it type checked and lint checked like everything else, and let someone test it without pushing to `main`. That is the change I would most like to see here.

Two smaller notes while you are in there: the script pages through every closed pull request on the repo, 251 and growing, on every run, when `sort=updated&direction=desc` plus an early break would do. And `datetime.datetime.utcnow()` is deprecated from Python 3.12, so `datetime.now(datetime.UTC)` if this stays Python.

The docs edits read well and I would keep them close to as written once the trigger question is settled.

File: .github/workflows/auto-release.yml [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        with:
+          fetch-depth: 0
+
+      - name: Configure git
+        run: |
+          git config user.name "github-actions[bot]"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+
+      - name: Setup Bun for scripts
+        uses: oven-sh/setup-bun@v2
```
[the remark this answers is not in the document]
Comment (Andes-indica) [reply .github/workflows/auto-release.yml]: resolved

File: .github/workflows/auto-release.yml [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+          if not breaking and not areas:
+            body += 'No merged pull requests found since the last tag.\n'
+
+          with open('RELEASE_NOTES.md','w') as f:
+            f.write(body)
+          print('WROTE RELEASE_NOTES.md')
+          # write output using the environment file instead of deprecated set-output
+          # the step runner will append this to $GITHUB_OUTPUT
+          PY
+          echo "notes_path=RELEASE_NOTES.md" >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (Andes-indica) [reply .github/workflows/auto-release.yml]: fixed

**Review by imshashank — CHANGES_REQUESTED** [review]
Good round. Five of the six things from the last review are done, and done properly:

- Push trigger gone, so this is weekly plus `workflow_dispatch` rather than a release per merge.
- `cancel-in-progress: false`, so a run can no longer be severed between pushing the tag and creating the release.
- Python replaced by `scripts/release-notes.ts` run with Bun, which is what the repo asks for.
- `gh release create` instead of raw `curl`, so a failed release actually fails the step.
- `oven-sh/setup-bun` pinned to a SHA alongside `actions/checkout`.

The script itself is a real improvement: typed, paginated with an early exit, and it fails loudly.

Two things are wrong, and the second one is my fault.

## 1. The since date step was deleted, not wired up

I asked you to pass `since_date` into the script's `env:` block. You did:

```yaml
env:
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
  since_date: ${{ steps.since.outputs.since_date }}
```

But the `Determine last tag and since` step that carried `id: since` is gone from the workflow. There is no step with that id any more, and GitHub resolves a reference to a missing step as an empty string rather than failing. So `since_date` is always empty and the script always takes its seven day fallback.

On a weekly cron that happens to be roughly right, which is why it will look fine in testing. It stops being right the moment a scheduled run is missed or delayed, or anyone triggers it by hand: the window is always the last seven days rather than "since the last tag", so merged pull requests can silently fall out of the notes entirely.

Either bring the step back and keep the wiring, or drop the env var and derive the date inside the script from `git describe --tags --abbrev=0`, which is probably cleaner now that the logic lives in TypeScript.

## 2. The early exit is unsound, and I suggested it

I told you to use `sort=updated&direction=desc` with an early break. That advice was wrong, and combined with this loop it drops releases on the floor:

```ts
const pagePRs = await fetchPageOfPRs(page);   // sorted by updated
for (const p of pagePRs) {
  if (!p.merged_at) continue;
  const mergedAt = new Date(p.merged_at);
  if (mergedAt < sinceDate) return merged;    // <- bails on the whole run
}
```

Sorting by `updated` does not order by `merged_at`. Someone comments on a pull request merged two months ago, its `updated_at` becomes today, and it sorts to the top of page one. Its `merged_at` is older than `since`, so this `return` fires on the first item and the release notes come back **empty** despite a week of merges.

That is not a rare shape. People comment on old pull requests constantly, and this repo has 257 closed ones to draw from.

The sound version keeps `sort=updated` for the paging bound, because a merged pull request's `updated_at` is always at least its `merged_at`, and stops filtering on the wrong field:

```ts
for (const p of pagePRs) {
  if (new Date(p.updated_at) < sinceDate) return merged;  // safe stop condition
  if (!p.merged_at) continue;
  if (new Date(p.merged_at) >= sinceDate) merged.push(p); // skip, do not stop
}
```

Sorry for sending you down that path.

## Smaller things

The docs in both files still say tags are published "weekly and on merges to `main`". The push trigger is gone, so that sentence now describes something the workflow does not do. Worth a pass over both paragraphs.

`scripts/release-notes.ts` has no test. That was fair enough when nothing under `scripts/` was tested, but #341 is adding `bun test scripts` to the root `test` script, so once that lands a test here will actually run in CI. The grouping and the date filtering are the parts worth pinning, and case 2 above is exactly the kind of thing a test catches.

Minor: `_githubSha` is prefixed like an unused binding but it is used in `renderNotes`. Drop the underscore.

File: .github/workflows/auto-release.yml [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        id: build_notes
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          
+        run: |
+          set -euo pipefail
+          bun ./scripts/release-notes.ts
+          echo "notes_path=RELEASE_NOTES.md" >> "$GITHUB_OUTPUT"
+
+      - name: Create dated tag
```
Comment (imshashank) [inline .github/workflows/auto-release.yml]: The tag is pushed before the release is created. If `gh release create` fails, retry sees the orphan, chooses a suffixed tag, and then uses the orphan as its cutoff, permanently dropping the failed interval. Derive the boundary from the latest successfully published automated release and make retry reuse or repair a same-SHA orphan.

File: scripts/release-notes.ts [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+    if (error.trim()) {
+      console.warn(`Could not determine last tag: ${error.trim()}`);
+    }
+    return null;
+  }
+
+  return output.trim() || null;
+}
+
+async function fetchPageOfPRs(page: number): Promise<PullRequest[]> {
```
Comment (imshashank) [inline scripts/release-notes.ts]: A wall-clock API query is not tied to the commit being released. A PR merged after this run starts can be advertised even though it is absent from `$GITHUB_SHA`, or can miss this fetch and then be excluded forever by the new tagger timestamp. PRs merged to another base can also enter. Generate notes from the exact Git commit range, then use GitHub data only to enrich that bounded set.

File: .github/workflows/auto-release.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+permissions:
+  contents: write
+  pull-requests: read
+
+jobs:
+  tag-and-release:
+    name: Create dated tag and release notes
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout
```
Comment (imshashank) [inline .github/workflows/auto-release.yml]: For `workflow_dispatch`, checkout follows the ref selected in the Actions UI, and the job later tags that `$GITHUB_SHA`. Reject non-main refs or explicitly fetch, resolve, and tag current `origin/main` so a manual run cannot publish a feature-branch commit.

**Review by imshashank — CHANGES_REQUESTED** [review]
Thanks for the follow-up. The Bun migration, full-SHA action pins, queued concurrency, visible release failures, annotated tagger timestamp, and corrected pagination boundary are good improvements.

I still need changes before this workflow is safe to enable:

1. Make publication idempotent and recover an existing tag without a release. A failed `gh release create` currently leaves an orphan, and retry skips it, uses it as the next cutoff, and permanently loses the failed release interval.
2. Build notes from the exact commit range being tagged, from the last successfully published automated release target through the current `main` SHA. The wall-clock query can advertise merges absent from the tag, omit merges forever, and include PRs merged to another base branch.
3. Ensure manual dispatch can publish only current `main`, not the selected workflow ref.
4. Select only successfully published dated releases as boundaries. `git describe --tags` accepts unrelated tags, and a reachable lightweight tag currently aborts the job.
5. Parse GitHub API responses with a shared Zod schema and add tests for cutoff boundaries, pagination, orphan recovery, same-day retries, grouping, and breaking-change guidance.
6. Correct the release documentation and formatting defects in the open docs thread.

Please also merge current `main`, rerun all checks, clear the Vercel failure, and reply to and resolve all five open review threads. The CodeQL file-write thread appears to be a false positive because the output path is fixed and the Markdown content is never executed, so a documented disagreement is sufficient there.

File: scripts/release-notes.ts [hunk 6]
```diff
@@ -0,0 +1,305 @@
...
+    if (releases.length < 100) break;
+  }
+
+  const latest = selectLatestPublishedRelease(published);
+
+  if (latest) {
+    return await runGit(['rev-list', '-n', '1', `${latest.tag_name}^{commit}`]);
+  }
+
+  return await runGit(['rev-list', '--max-parents=0', targetSha]);
```
[the remark this answers is not in the document]
Comment (Andes-indica) [reply scripts/release-notes.ts]: resolved it was a good catch

**Comment by Andes-indica** [comment]
Addressed in the latest update. The comment was valid.

Tag selection now distinguishes three cases:

* an existing tag at the target SHA is reused;
* an unpublished orphan at an older SHA is repaired to the current target;
* a tag that already has a published release is preserved, and a suffixed tag is selected.

The workflow rechecks that the orphan has no GitHub release before repairing it and updates the tag using `--force-with-lease` to avoid overwriting a concurrent change. The release boundary continues to come only from the latest successfully published automated release.

I also added tests for advanced-main orphan repair and for preserving already-published tags. The focused release-note tests, typecheck, and full verification pass.

if any thing i am missing please update me i will work on it

**Comment by imshashank** [comment]
Thanks for the update and the screenshots confirming the manual run. Latest commit (`fa29954`) looks like it addresses the outstanding items from the last review: idempotent publication with orphan-tag recovery, notes built from the exact tagged commit range, manual dispatch pinned to `main`, boundary selection restricted to successfully published automated releases, and Zod validation with tests.

This is queued for a full re-review pass, not stalled on you. I'll follow up here once I've been through the current head.

**Comment by Andes-indica** [comment]
Thanks for confirming. I’ll wait for the full re-review.

**Comment by imshashank** [comment]
Status update after `7ecc729b`: the branch now includes current `main` and the repository-owned CI workflow is green. The merge did not change the release implementation, however, so the blockers from the full review remain: `reuse_tag` is still read as `tag_action`, the publish step still receives no `TAG`, and the repair lookup still runs without `GH_TOKEN` and can fail open before a force-move. Checkout also retains the write credential while repository code runs.

I submitted a fresh changes-requested review anchored to the current head. The PR still has 12 unresolved threads, CodeRabbit is paused, and Greptile is 3/5. It continues to wait on the author, not on a maintainer merge.

File: .github/workflows/auto-release.yml [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+              EXISTING_SHA=$(git rev-list -n 1 "$TAG^{commit}")
+              if [ "$EXISTING_SHA" != "$TARGET_SHA" ]; then
+                echo "Selected tag $TAG does not point to target $TARGET_SHA"
+                exit 1
+              fi
+              echo "Reusing existing tag $TAG at $TARGET_SHA"
+              ;;
+
+            repair)
+              if gh release view "$TAG" --repo "$GITHUB_REPOSITORY" >/dev/null 2>&1; then
```
Comment (imshashank) [inline .github/workflows/auto-release.yml]: [P1] This step does not provide `GH_TOKEN`, and checkout's persisted Git credential does not authenticate `gh`. More importantly, every `gh release view` failure is interpreted as no release, so an auth, rate-limit, or API failure can enter the repair branch and force-move a tag that already has a published release. Provide a step-scoped token and distinguish an explicit not-found response from every other failure. Only an explicit 404 may permit repair.

**Review by imshashank — CHANGES_REQUESTED** [review]
Current-head re-review on `fa29954a`: pinning manual runs to current `main`, selecting the last successfully published dated release, generating notes from the exact commit range, and modeling orphan recovery are good improvements.

The workflow still cannot publish. The script emits `reuse_tag` while the workflow reads `tag_action`; the selected tag is not exported or passed into the publish step; and the repair check runs `gh` without a token while treating every lookup failure as permission to force-move the tag. Please also address the open credential-persistence, request-timeout, test-ownership, and documentation threads, and add a workflow contract test that catches the output and step-environment wiring.

The branch is 9 commits behind `main`, has 11 unresolved threads, a failed Vercel status, and no approval. Please fix these, merge current `main`, rerun all checks, and clear every thread. This is not safe to merge.

Ratings: security 1/5, correctness 1/5, performance 2/5, maintainability 2/5, tests 2/5.

**Review by imshashank — CHANGES_REQUESTED** [review]
Re-reviewed current head `7ecc729b`. Merging current `main` made CI green and removed the behind-main condition, but it did not change the release workflow wiring, so the functional blockers remain:

1. `scripts/release-notes.ts` writes the action as `reuse_tag`, while the workflow reads `steps.build_notes.outputs.tag_action`. `TAG_ACTION` is therefore empty and the case statement exits through `Unknown tag action`.
2. The publish step reads `$TAG`, but that step receives no `TAG` environment value and `release_tag` never writes a tag output. With `set -u`, publication stops on an unbound variable even after the first mismatch is fixed.
3. The repair path calls `gh release view` without `GH_TOKEN` and treats every nonzero result as proof that no release exists. Authentication, rate-limit, and API failures must fail closed; only an authenticated not-found result may permit a force-move.
4. Checkout still leaves the write credential persisted while dependency installation and the repository script run. Set `persist-credentials: false` and scope authenticated Git operations to the mutation boundary.
5. Add an executable workflow-contract test that locks the script output names, cross-step tag propagation, authenticated repair lookup, and failure behavior. The current release-note unit tests cannot catch these workflow failures.

There are also 12 unresolved review threads. CodeRabbit remains paused and the current Greptile review is 3/5. This head must not merge yet.

File: .github/workflows/auto-release.yml [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+                  ;;
+              esac
+
+              OLD_TAG_OBJECT=$(git rev-parse "refs/tags/$TAG")
+              echo "Repairing unpublished tag $TAG at $TARGET_SHA"
+              git tag -fa "$TAG" "$TARGET_SHA" -m "Automated release $TAG"
+              git push \
+                --force-with-lease="refs/tags/$TAG:$OLD_TAG_OBJECT" \
+                "$REPOSITORY_URL" \
+                "refs/tags/$TAG"
```
[the remark this answers is not in the document]
Comment (Andes-indica) [reply .github/workflows/auto-release.yml]: Good catch the race is valid. The ref lease only protects the tag itself and cannot prevent a release from being created between the 404 response and the force-push.

I propose removing tag force-moves entirely. When an unpublished orphan points to an older commit, the recovery run will keep the tag at that commit, generate and publish the missing release for the exact previous-release-to-orphan range, and leave newer commits for the following run. For example, with published boundary A, orphan B, and current main C, recovery publishes A..B first; the next run publishes B..C.

This keeps tags immutable, preserves exact release ranges, and removes the check-then-mutate race. Does that recovery model match the behavior you expect before I refactor the workflow, generator, and contract tests?

**Review by imshashank — CHANGES_REQUESTED** [review]
Re-reviewed current head `9cb3df4`. I pulled `.github/workflows/auto-release.yml` directly rather than going off the summary, and this head is not closer to working than `7585b19` was, it's regressed.

Three things Greptile flagged as P1 are real, I checked each against the raw file:

1. **The `Select or recover dated tag` step is not a step.** The preceding `Build release notes` step ends its `run: |` block, and the next line, `            - name: Select or recover dated tag`, is indented at 12 spaces, deeper than the `run:` key at 8. YAML keeps consuming that block scalar, so this entire step is swallowed as literal text inside the previous step's shell script. There is no `id: release_tag` step in this workflow as currently indented. Every later reference to `steps.release_tag.outputs.tag` resolves to nothing.
2. **`RELEASE_TARGET_SHA` self-references.** `Build release notes` sets `env: RELEASE_TARGET_SHA: ${{ steps.build_notes.outputs.release_target_sha }}`, reading its own step's not-yet-produced output instead of `steps.target.outputs.target_sha` from the preceding `Determine release target` step. This is always empty.
3. **`TAG_ACTION` is still disconnected**, the same finding from my Aug 23 review on `7ecc729b`. The step reads `steps.build_notes.outputs.tag_action`, and Greptile confirms the script emits `reuse_tag`, not `tag_action`. Given point 1, this step doesn't even execute as a separate step right now, so this is moot until the indentation is fixed, but it will resurface immediately once it is.

One more I didn't see flagged anywhere yet: the final `Publish or recover GitHub release` step sets `env: GH_TOKEN: ${{ secrets.target.outputs.target_sha }}`. That mixes the `secrets` context with a `steps.*.outputs` path, it isn't a valid reference to anything, so `GH_TOKEN` is empty and every `gh` call in that step fails auth. That step needs `GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}`, matching what the tag-selection step already does correctly a few lines up.

Net effect: as written, a scheduled or manually dispatched run cannot produce a tag or a release. `checkout` with `persist-credentials: false` and `cancel-in-progress: false` from earlier rounds are in and correct, so this isn't a case of nothing landing, but the core wiring is broken in a new way each round. I'd suggest actually running this workflow via `workflow_dispatch` on a fork before the next push, a contract test that checks string content (like the current `tests/auto-release-workflow.test.ts`) can't catch a block-scalar indentation bug, only parsing the YAML and asserting on the resulting step list would.

10 unresolved review threads remain, this doesn't add a new one since points 1-3 already have open threads, I'm confirming them rather than opening duplicates. Retaining `waiting-for-author`, this must not merge in its current state.

**Comment by imshashank** [comment]
Status check on the new push (`a3021d3c`), not stalled on you anymore.

I read the workflow and the script directly rather than the summary. The three wiring bugs from my last review are actually fixed this time:

- `Build release notes` now reads `RELEASE_TARGET_SHA` from `steps.target.outputs.target_sha` instead of self-referencing its own not-yet-produced output.
- `Select or recover dated tag` is a real, correctly indented step again (`id: release_tag`), and its env now lines up with what the script emits: `TAG_ACTION` reads `steps.build_notes.outputs.tag_action`, matching `selectedTag.action` (`create` / `reuse` / `recover`).
- `Publish or recover GitHub release` now sets `GH_TOKEN: ${{ secrets.GITHUB_TOKEN }}` instead of the mixed `secrets`/`steps` reference that broke every `gh` call.

Exact-head CI is fully green: lint/comments/types, unit/integration, Playwright, build, migrations, CodeQL, TypeScript analysis, Markdown links, and Greptile. Only the expected external MagicAPI Vercel authorization gate is red.

Two things before I can approve:

1. There are 30 open review threads on this PR. A lot of them are almost certainly stale, tied to commits several pushes back, but I'm not going to assume that without checking each one. Please reply to or resolve the ones that are actually addressed by this head, and leave the rest open with a note if you disagree.
2. I want to re-read the `recover` path once more end to end now that it no longer force-moves the tag, just to confirm the release notes range and the recovered tag's target stay consistent. That's on me, not you.

Swapping `waiting-for-author` for `needs-review`. Nothing further needed from you until I've been through the threads, unless you want to start closing out the ones you know are already fixed.

File: .github/workflows/auto-release.yml [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+            SINCE_DATE=$(date -u -d '7 days ago' --iso-8601=seconds)
+            echo "No tag found, using fallback since: $SINCE_DATE"
+          fi
+          echo "last_tag=$LAST_TAG" >> "$GITHUB_OUTPUT"
+          echo "since_date=$SINCE_DATE" >> "$GITHUB_OUTPUT"
+
+      - name: Build release notes (Python)
+        id: build_notes
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
[the remark this answers is not in the document]
Comment (Andes-indica) [reply .github/workflows/auto-release.yml:52]: resolved in a3021d3c

File: .github/workflows/auto-release.yml [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+
+on:
+  push:
+    branches: [main]
+  schedule:
+    - cron: '0 0 * * 0'
+  workflow_dispatch:
+concurrency:
+  group: automated-release
+  cancel-in-progress: true
```
[the remark this answers is not in the document]
Comment (Andes-indica) [reply .github/workflows/auto-release.yml]: resolved in a3021d3c

File: .github/workflows/auto-release.yml (hunk 3 above)
Comment (Andes-indica) [reply .github/workflows/auto-release.yml]: resolved in a3021d3c

File: scripts/release-notes.ts (hunk 4 above)
Comment (Andes-indica) [reply scripts/release-notes.ts]: resolved in a3021d3c

File: scripts/release-notes.ts [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+  }
+
+  const checkedOutSha = await runGit(['rev-parse', 'HEAD']);
+  if (checkedOutSha !== targetSha) {
+    throw new Error(`Checked-out main is ${checkedOutSha}, expected release target ${targetSha}`);
+  }
+
+  const published = await getPublishedReleases();
+  const publishedTags = new Set(published.map((release) => release.tag_name));
+  const baseTag = new Date().toISOString().slice(0, 10).replaceAll('-', '.');
```
[the remark this answers is not in the document]
Comment (imshashank) [reply scripts/release-notes.ts:370]: Fixed in `1da4666a`. The generator now scans all dated tags, validates the first-parent path from the last published boundary to the current target, and recovers the earliest unpublished tag on that eligible path before allocating the current date. Tags before the boundary or off main remain immutable and are ignored, and an occupied current-date name advances to a suffix. Regressions cover cross-date recovery, consecutive ranges, multiple orphans, off-main and pre-boundary tags, first-release root recovery, and an invalid boundary. The 27 focused tests, lint, comments check, and typecheck pass locally.

**Comment by imshashank** [comment]
@coderabbitai review

File: scripts/release-notes.ts [hunk 12]
```diff
@@ -0,0 +1,408 @@
...
+export function selectDatedTag(
+  baseTag: string,
+  targetSha: string,
+  existingTags: Record<string, string>,
+  publishedTags: Set<string>,
+  eligibleCommits: readonly string[] = [],
+): { tag: string; action: TagAction; releaseTargetSha: string } {
+  const commitOrder = new Map(eligibleCommits.map((sha, index) => [sha, index]));
+  const orphan = Object.entries(existingTags)
+    .filter(([tag, sha]) => !publishedTags.has(tag) && commitOrder.has(sha))
```
[the remark this answers is not in the document]
Comment (imshashank) [reply scripts/release-notes.ts]: Fixed in c4c6d60a. Tag selection now searches the complete target first-parent history, so it recovers the earliest unpublished dated tag even when a newer tag already has a release. Release-note boundaries are selected independently by commit order: a historical recovery uses the nearest older published commit, and the following current release uses the nearest newer published commit rather than publication timestamps. Off-main tags are ignored and invalid targets fail closed. Regressions cover the stranded-orphan scenario, the recovery range, the subsequent current range, root fallback, and off-main tags. The 31 focused tests, lint, comment policy, all workspace typechecks, and diff checks pass.

File: scripts/release-notes.ts [hunk 13]
```diff
@@ -0,0 +1,391 @@
...
+export function selectDatedTag(
+  baseTag: string,
+  targetSha: string,
+  existingTags: Readonly<Record<string, string>>,
+  publishedTags: ReadonlySet<string>,
+  firstParentHistory: readonly string[] = [],
+): { tag: string; action: TagAction; releaseTargetSha: string } {
+  const commitOrder = new Map([...firstParentHistory].reverse().map((sha, index) => [sha, index]));
+  const orphan = Object.entries(existingTags)
+    .filter(([tag, sha]) => !publishedTags.has(tag) && commitOrder.has(sha))
```
[the remark this answers is not in the document]
Comment (imshashank) [reply scripts/release-notes.ts:184]: Fixed in 1e0fa1e8. If a published dated tag is newer than the selected historical orphan on the same first-parent path, the recovery now uses the orphan target itself as the boundary. That produces an empty historical range and cannot duplicate changes already covered by the newer release. Without a newer covering release, boundary selection still uses the nearest older published commit or the root. Regressions cover newer-only and newer-plus-older boundaries, multiple historical orphan recoveries with no overlap, off-main tags, and the following current release returning to the newest published commit. The 33 focused tests, lint, comment policy, all workspace typechecks, diff checks, and push hooks pass.

**Review by imshashank — APPROVED** [review]
Exact-head approval for 1e0fa1e8.

The release workflow and generator are current with main. Historical orphan recovery selects the earliest unpublished dated tag on first-parent history, keeps existing tags immutable, uses commit order for release boundaries, and prevents overlapping recovery notes. Off-history tags are ignored.

Focused verification is 33 passing tests. Hosted lint, policy, types, unit and integration tests, Playwright, build, migrations, links, CodeQL, CodeRabbit, Greptile, and Vercel all pass on this exact head. CodeRabbit reports Review completed, Greptile is 5/5, and all 33 review threads are resolved. The temporary internal Vercel validation ref was deleted after the exact-SHA deployment succeeded.

No review blocker remains.

**Comment by Andes-indica** [comment]
Thank you for the detailed reviews and guidance throughout this PR. I learned a lot while working through the workflow reliability, recovery, security, and testing concerns. Glad to see it merged!
